### PR TITLE
Reduce allocations during typing

### PR DESCRIPTION
**Pull request by ToddGrun**
During typing, various codepaths attempt to find the set of changes between two text images. Note that while this change will get rid of the allocations, I'm not certain it makes sense for that codepath to be getting executed so frequently and am following up elsewhere. This should reduce allocations in the profile that I'm looking at by around 1.4% during typing.

*** allocations from the typing scenario in the new csharp editing speedometer that should be alleviated **
image

### CODE REVIEW

File: src/EditorFeatures/Text/ITextImageHelpers.cs [hunk 1]
```diff
@@ -51,35 +52,54 @@ public static IReadOnlyList<TextChangeRange> GetChangeRanges(ITextImageVersion o
         }
 
         if (changes == null)
-        {
             return [];
-        }
-        else
+
+        var builder = ArrayBuilder<TextChangeRange>.GetInstance(changes.Count);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Text/ITextImageHelpers.cs:57]: FixedArrayBuilder?

File: src/EditorFeatures/Text/ITextImageHelpers.cs [hunk 2]
```diff
@@ -51,35 +52,54 @@ public static IReadOnlyList<TextChangeRange> GetChangeRanges(ITextImageVersion o
...
         TextChangeRange? range = null;
-        var iterator = GetMultipleVersionTextChanges(oldVersion, newVersion, forward);
-        foreach (var changes in forward ? iterator : iterator.Reverse())
+        using var _ = ArrayBuilder<ArrayBuilder<TextChangeRange>>.GetInstance(out var builder);
+
+        GetMultipleVersionTextChanges(oldVersion, newVersion, forward, builder);
+        foreach (var changes in builder)
         {
             range = range.Accumulate(changes);
+            changes.Free();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Text/ITextImageHelpers.cs:76]: nit.  consider moving into a try/finally.

File: src/EditorFeatures/Text/ITextImageHelpers.cs [hunk 3]
```diff
@@ -51,35 +52,54 @@ public static IReadOnlyList<TextChangeRange> GetChangeRanges(ITextImageVersion o
...
+    private static void GetMultipleVersionTextChanges(
+        ITextImageVersion oldVersion,
+        ITextImageVersion newVersion,
+        bool forward,
+        ArrayBuilder<ArrayBuilder<TextChangeRange>> builder)
     {
         for (var version = oldVersion; version != newVersion; version = version.Next)
         {
-            yield return version.Changes.Select(forward ? s_forwardTextChangeRange : s_backwardTextChangeRange);
+            var changes = ArrayBuilder<TextChangeRange>.GetInstance(version.Changes.Count);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Text/ITextImageHelpers.cs:91]: FixedArrayBuilder?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/EditorFeatures/Text/ITextImageHelpers.cs (hunk 1 above)
Comment (ToddGrun) [reply src/EditorFeatures/Text/ITextImageHelpers.cs:57]: Will do!

File: src/EditorFeatures/Text/ITextImageHelpers.cs (hunk 3 above)
Comment (ToddGrun) [reply src/EditorFeatures/Text/ITextImageHelpers.cs:91]: Gonna skip on this one as it's easier to just use the ArrayBuilder as it's passed out of the method.

File: src/EditorFeatures/Text/ITextImageHelpers.cs (hunk 2 above)
Comment (ToddGrun) [reply src/EditorFeatures/Text/ITextImageHelpers.cs:76]: Unless you feel strongly, I'd rather keep as. Exceptions shouldn't be happening during the Accumulate call, and it they do, not putting a couple arrays back in the pool is probably ok.

**Review by 333fred — APPROVED** [review]
Compiler change LGTM (commit 2)

File: src/EditorFeatures/Text/ITextImageHelpers.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Text/ITextImageHelpers.cs:76]: i do not feel strongly here.  this is simple enough :)

**Comment by ToddGrun** [comment]
@dotnet/roslyn-compiler for 2nd review

File: src/Compilers/Core/Portable/Text/TextChangeRangeExtensions.cs [hunk 4]
```diff
@@ -25,7 +24,9 @@ internal static class TextChangeRangeExtensions
             // we could apply each one individually like we do in SyntaxDiff::ComputeSpansInNew by calculating delta
             // between each change in changesInNextVersion which is already sorted in its textual position ascending order.
             // but end result will be same as just applying it once with encompassed text change range.
-            var newChange = TextChangeRange.Collapse(changesInNextVersion);
+            var newChange = changesInNextVersion.Count == 1
+                ? changesInNextVersion[0]
+                : TextChangeRange.Collapse(changesInNextVersion);
```
Comment (Youssef1313) [inline src/Compilers/Core/Portable/Text/TextChangeRangeExtensions.cs:29]: Should this optimization be moved inside `TextChangeRange.Collapse` itself?

Comment (ToddGrun) [reply src/Compilers/Core/Portable/Text/TextChangeRangeExtensions.cs:29]: I considered that, but that method has several usages in which the caller only has an IEnumerable. I instead chose to limit the changes to the caller that was showing up in the performance profile I was looking at.

**Review by jjonescz — APPROVED** [review]

**Pull request merged**